src/program.cpp

```diff
@@ -113,342 +86,115 @@ program& program::operator=(program p)
-    assert(std::all_of(first, last, [&](const instruction& ins) { return ins.outputs().empty(); }));
-    return impl->instructions.erase(first, last);
-}
-
-instruction_ref program::move_instruction(instruction_ref src, instruction_ref dst)
-{
-    impl->instructions.splice(dst, impl->instructions, src);
-    return src;
-}
-
-instruction_ref program::move_instructions(instruction_ref src, instruction_ref dst)
-{
-    this->move_instruction(src, dst);
-    for(auto ins : src->inputs())
-        this->move_instruction(ins, src);
-    return src;
-}
-
-instruction_ref program::add_literal(literal l)
+shape program::get_parameter_shape(std::string name) const
 {
-    impl->instructions.emplace_front(std::move(l));
-    return impl->instructions.begin();
+    const auto* mm = this->get_main_module();
+    return mm->get_parameter_shape(std::move(name));
 }
 
-instruction_ref program::add_outline(const shape& s)
+std::vector<std::string> program::get_parameter_names() const
 {
-    impl->instructions.push_front({builtin::outline{s}, s, {}});
-    return impl->instructions.begin();
+    const auto* mm = this->get_main_module();
+    return mm->get_parameter_names();
 }
 
-instruction_ref program::add_parameter(std::string name, shape s)
+instruction_ref program::get_parameter(std::string name) const
 {
-    assert(get_parameter_shape(name) == shape{});
-    impl->input_names.push_back(name);
-
-    impl->instructions.push_front({builtin::param{std::move(name)}, std::move(s), {}});
-    return impl->instructions.begin();
+    const auto* mm = this->get_main_module();
+    return mm->get_parameter(std::move(name));
 }
 
-instruction_ref program::add_return(std::vector<instruction_ref> args)
+std::unordered_map<std::string, shape> program::get_parameter_shapes() const
 {
-    assert(std::all_of(
-               args.begin(), args.end(), [&](instruction_ref x) { return has_instruction(x); }) &&
-           "Argument is not an exisiting instruction");
-    impl->instructions.push_back({builtin::returns{}, {}, args});
-    auto result = std::prev(impl->instructions.end());
-    instruction::backreference(result);
-    assert(result->valid(begin()));
-
-    return result;
+    const auto* mm = this->get_main_module();
+    return mm->get_parameter_shapes();
 }
 
-shape program::get_parameter_shape(std::string name) const
-{
-    auto ins = std::find_if(
-        impl->instructions.begin(), impl->instructions.end(), [&](const instruction& x) {
-            if(x.name() == "@param")
-            {
-                return any_cast<builtin::param>(x.get_operator()).parameter == name;
-            }
-            else
-            {
-                return false;
-            }
-        });
-    if(ins != this->end())
-        return ins->get_shape();
-    else
-        return {};
-}
-
-std::vector<std::string> program::get_parameter_names() const
+bool program::has_instruction(instruction_ref ins) const
```

Comment:
This method should only be in the module.

Reply:
If a program does not have a has_instruction(), how can we know whether a program contains an instruction? Use the approach of going through all modules to do the check?

Reply:
Do we call `has_instruction` on a program? I think we just use this over modules. If we do need program-wide `has_instruction` then it should go through all of the modules.

Reply:
removed